homeassistant/components/hangouts/hangouts_bot.py

```diff
@@ -0,0 +1,228 @@
+
+    async def async_disconnect(self):
+        """Disconnect the client if it is connected."""
+        if self._connected:
+            await self._client.disconnect()
+
+    async def async_handle_hass_stop(self, _):
+        """Run once when Home Assistant stops."""
+        await self.async_disconnect()
+
+    async def _async_send_message(self, message, targets):
+        conversations = []
+        for target in targets:
+            conversation = None
+            if 'id' in target:
+                conversation = self._conversation_list.get(target['id'])
+            elif 'name' in target:
+                conversation = self._resolve_conversation_name(target['name'])
+            if conversation is not None:
+                conversations.append(conversation)
+
+        if not conversations:
+            return False
+
+        from hangups import ChatMessageSegment, hangouts_pb2
+        messages = []
+        for segment in message:
+            if 'parse_str' in segment and segment['parse_str']:
+                messages.extend(ChatMessageSegment.from_str(segment['text']))
+            else:
+                if 'parse_str' in segment:
+                    del segment['parse_str']
+                messages.append(ChatMessageSegment(**segment))
+            messages.append(ChatMessageSegment('',
+                                               segment_type=hangouts_pb2.
+                                               SEGMENT_TYPE_LINE_BREAK))
+
+        if not messages:
+            return False
+        for conv in conversations:
+            await conv.send_message(messages)
+
+    async def _async_list_conversations(self):
+        import hangups
+        self._user_list, self._conversation_list = \
+            (await hangups.build_user_conversation_list(self._client))
+        users = {}
+        conversations = {}
+        for user in self._user_list.get_all():
+            users[str(user.id_.chat_id)] = {'full_name': user.full_name,
+                                            'is_self': user.is_self}
+
+        for conv in self._conversation_list.get_all():
+            users_in_conversation = {}
+            for user in conv.users:
+                users_in_conversation[str(user.id_.chat_id)] = \
+                    {'full_name': user.full_name, 'is_self': user.is_self}
+            conversations[str(conv.id_)] = \
+                {'name': conv.name, 'users': users_in_conversation}
+
+        self.hass.states.async_set("{}.users".format(DOMAIN),
+                                   len(self._user_list.get_all()),
+                                   attributes=users)
+        self.hass.bus.fire(EVENT_HANGOUTS_USERS_CHANGED, users)
+        self.hass.states.async_set("{}.conversations".format(DOMAIN),
+                                   len(self._conversation_list.get_all()),
+                                   attributes=conversations)
+        self.hass.bus.fire(EVENT_HANGOUTS_CONVERSATIONS_CHANGED, conversations)
```

Comment:
This should be `async_fire` if you want to fire events. However, is this for internal communication of the component or is this for people to automate on? Only for the latter should it be an event, use dispatcher otherwise.